compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.h

```diff
@@ -45,13 +47,15 @@ class VISQErrorApproximator final
 
 private:
   /**
-   * @brief initiliaze by visq_data (returns success)
+   * @brief initiliaze by visq_data (throws on invalid input)
    */
-  bool init(std::istream &visq_data);
+  void init(std::istream &visq_data);
 
 private:
   std::string _visq_data_path;
   std::map<std::string, float> _layer_errors;
+
+  friend CircleMPQSolverVISQErrorApproximatorTest;
```

Comment:
Adding this ~Test class here looks odd.
Do you know any practice doing so?

Reply:
@seanshpark 
It's added to the class for unitests. For 'ONE` repository `friend` are added to the top of the class. Am i supposed to add it to the top?

Reply:
I know the usage but this class or any other class doesn't have any relation with `CircleMPQSolverVISQErrorApproximatorTest`. 
`CircleMPQSolverVISQErrorApproximatorTest` is only for testing and testing is not included in any production.

And I don't agree on usging `~Test` as a friend of a class.

Reply:
Are tests supposed to use `temporary` files?

Reply:
> Will it be ok if tests will use temporary files?

Our tests uses lots of file based testing. It is OK to use files.
I assume you choose to use memory (stream) testing.
Was there any assumption behind this? like does it consume huge space? or long time?

Alternative thing I could think of is to use some name like `ErrorApproximatorAccessor` class, not having `Test`, but as I asked above, I'd like to know any know practice like this.

Reply:
> Alternative thing I could think of is to use some name like `ErrorApproximatorAccessor` class, not having `Test`, but as I asked above, I'd like to know any know practice like this.

No. I've not seen this in `ONE`.
I'd rename it to `VISQErrorApproximatorAccessor`.

Reply:
> Was there any assumption behind this?

Yes. Like it's slower than memory testing.

Reply:
> No. I've not seen this in ONE.

Have you seen this in any other projects? or is this your invention?

Reply:
I've seen it here:
https://stackoverflow.com/questions/47354280/what-is-the-best-way-of-testing-private-methods-with-googletest

Reply:
> I recommend avoiding option 1 and 2 when you can, as it typically causes "test duplication", and as a consequence, many more tests than necessary will break when requirements change. Use them as a last resort. Option 1 and 2 are the fastest ways to "test private methods" for here and now (as in the fastest to implement), but they'll really hurt productivity in the long run.

which recommends not using `friend` class for testing.

option 1/ use with file testing with some comments as you worry about performance.
option 2/ use `protected` for `init(stream)` and inherit this class in Test.